cc/model/features.h

```diff
@@ -44,52 +47,159 @@ namespace minigo {
-  MG_ALWAYS_INLINE static void Set(const ModelInput& input, int num_planes,
-                                   T* dst) {
+  MG_ALWAYS_INLINE static void SetNhwc(const ModelInput& input, int num_planes,
+                                       T* dst) {
     auto my_color = input.position_history[0]->to_play();
     auto their_color = OtherColor(my_color);
 
+    auto n = std::min(input.position_history.size(), PositionHistory);
+
     // Write the features for the position history that we have.
     int j = 0;
-    for (; j < input.position_history.size(); ++j) {
+    for (; j < n; ++j) {
       const auto* src = input.position_history[j]->stones().data();
       const auto* end = dst + kN * kN * num_planes;
       for (auto* d = dst + j * 2; d < end; d += num_planes) {
         auto color = src->color();
         src += 1;
-        d[0] = color == my_color ? 1 : 0;
-        d[1] = color == their_color ? 1 : 0;
+        d[0] = color == my_color;
+        d[1] = color == their_color;
       }
     }
 
     // Pad the features with zeros if we have fewer than 8 moves of history.
-    for (; j < kMaxPositionHistory; ++j) {
+    for (; j < PositionHistory; ++j) {
       const auto* end = dst + kN * kN * num_planes;
       for (auto* d = dst + j * 2; d < end; d += num_planes) {
         d[0] = 0;
         d[1] = 0;
       }
     }
   }
+
+  MG_ALWAYS_INLINE static void SetNchw(const ModelInput& input, float* dst) {
+    auto my_color = input.position_history[0]->to_play();
+    auto their_color = OtherColor(my_color);
+
+    auto n = std::min(input.position_history.size(), PositionHistory);
+
+    // Write the features for the position history that we have.
+    int j = 0;
+    for (; j < n; ++j) {
+      const auto& stones = input.position_history[j]->stones();
+      for (const auto& stone : stones) {
+        *dst++ = stone.color() == my_color;
+      }
+      for (const auto& stone : stones) {
+        *dst++ = stone.color() == their_color;
+      }
+    }
+
+    // Pad the features with zeros if we have fewer than 8 moves of history.
+    for (; j < PositionHistory; ++j) {
+      for (int i = 0; i < 2 * kN * kN; ++i) {
+        *dst++ = 0;
+      }
+    }
+  }
+
+  MG_ALWAYS_INLINE static void SetNchw(const ModelInput& input, uint8_t* dst) {
+    auto my_color = input.position_history[0]->to_play();
+    auto their_color = OtherColor(my_color);
+
+    auto n = std::min(input.position_history.size(), PositionHistory);
+
+    // Stones are stored as bit-packed uint16_t, with the color stored in the
+    // bottom two bits.
+    static_assert(sizeof(Stone) == 2, "");
+    __m128i color_mask = _mm_set1_epi16(3);
+
+    // Initialize useful single byte values.
+    __m128i one = _mm_set1_epi8(1);
+    __m128i my_color_mm = _mm_set1_epi8(static_cast<int>(my_color));
+    __m128i their_color_mm = _mm_set1_epi8(static_cast<int>(their_color));
+
+    // Write the features for the position history that we have.
+    int safe_size = (kN * kN / 16) * 16;
```

Comment:
consider adding
int layer_size = kN * kN;
to aid comprehension

Reply:
Good call. Added a `kNumPoints = kN * kN` constant to `cc/constants.h`. I should have done that a long time ago.